doc/sphinx/user-extensions/syntax-extensions.rst

```diff
@@ -1100,113 +1081,27 @@ special identifier called a *delimiting key* and bound to a given scope.
 In such a situation, the term term, and all its subterms, are
 interpreted in the scope stack extended with the scope bound to :token:`ident`.
 
-.. cmd:: Delimit Scope @scope with @ident
+.. cmd:: Delimit Scope @ident with @scope
```

Comment:
This change and your previous `todo` make me realize that we have be using the `scope` nonterminal inconsistently throughout the grammar. Sometimes to mean the scope name itself (as in `: @scope`) and sometimes to mean the delimiting key (as in `% @scope`). We could probably change this to `@scope_name` and `@scope_key`. Another solution would is to backtrack on the introduction of this nonterminal and to use instead `@ident__scopename` and `@ident__scopekey` when this is deemed useful and just `@ident` everywhere else.

Reply:
Let's use `@scope_name` and `@scope_key`.  Commands always take one or the other but never both?  One of them is always in the form `XXX_scope`?  Why is this a useful distinction?  If it is, we should explain why.  Can you identify which places take which type of scope?

> Another solution would is ... to use instead `@ident__scopename` and `@ident__scopekey` when this is deemed useful and just `@ident` everywhere else.

No, we should use subscripts only when we have no alternative.

Reply:
>  Commands always take one or the other but never both?

When a scope is passed as argument to a command, it is generally a scope name.
One command takes both, it is: `Delimit Scope @scope_name with @scope_key`.

>  One of them is always in the form `XXX_scope`?

It is just a convention of the standard library, but Coq does not enforce it.

> Why is this a useful distinction? If it is, we should explain why.

It is my impression that the doc already explains this distinction. Scopes must always have a name, but having a delimiting key is optional. When you only have the name, you can attach notations to a scope, open and close interpretation scopes. This is already pretty useful. When you have a delimiting key, you additionally have access to the `% @scope_key` syntax which gives you even more flexibility. Usually, the scope name is longer to be less ambiguous and the delimiting key is shorter to be fewer characters to type, but it is perfectly acceptable to use the same identifier for both the scope name and the delimiting key.

It is my impression however, that the `%` syntax in the `Arguments` command accepts *both* scope names or delimiting keys.

Reply:
> It is my impression however, that the % syntax in the Arguments command accepts both scope names or delimiting keys (oh right, this is already written in the doc of Arguments).

Yes, and see also #9123.

Reply:
> > Why is this a useful distinction?

Ah, I'm asking about the design, not Coq's current behavior.  Why shouldn't scope names and scope keys be interchangeable except where it obvious matters (e.g. defining a new scope).

Reply:
If they were fully interchangeable than you couldn't have a useful `Undelimit Scope` command to remove the printing of notations when the scope is closed.  Also, in theory you can have a single scope bound to multiple keys, and you could also have multiple scopes bound to the same key.  (Does Coq behave well in either of these cases?  I've never tested them.). If these uses are deemed not worth keeping, then `Undelimit Scope` could be replaced with `Disable Printing` or something and scope names and keys could be combined.

Reply:
Clearly `Delimit Scope` and `Undelimit Scope` should apply to the keys.  (You did see my qualification "except where it obviously matters"?)  OTOH, why shouldn't users be able to specify a scope name on a subterm rather than a scope key?

> If they were fully interchangeable than you couldn't have a useful Undelimit Scope command to remove the printing of notations when the scope is closed.

If that's a common use case, perhaps you would provide a simple example that I can add to the documentation?

What are the other use cases?  I don't know them.

> in theory you can have a single scope bound to multiple keys

That seems to be permitted but `Print Scopes` only shows one

> you could also have multiple scopes bound to the same key

If it's useful and I have an example I will mention it.

Reply:
> OTOH, why shouldn't users be able to specify a scope name on a subterm rather than a scope key?

Because if users can do it, then they might ask "why can't the printer do it" (parsing and printing should be somewhat symmetric, I think), and the reason printing shouldn't be able to do it is because then `Undelimit Scope` will no longer work for turning off printing of a scope.

> If it's useful and if I have an example I will mention it.

I'm not sure if it's useful, but I can make an example if you want

Reply:
> If they were fully interchangeable than you couldn't have a useful Undelimit Scope command to remove the printing of notations when the scope is closed.
>
> If that's a common use case, perhaps you would provide a simple example that I can add to the documentation?
>
> What are the other use cases? I don't know them.

```coq
Definition foo := 10.
Print foo. (* foo = 10 *)
Close Scope nat_scope.
Print foo. (* foo = 10%nat *)
Undelimit Scope nat_scope.
Print foo. (* foo = S 9 *)
Arguments S : clear scopes.
Print foo. (* foo = S (S (S (S (S (S (S (S (S (S O))))))))) *)
```

Reply:
An interesting example but not clear that it's useful in practice.  The last 4 lines of this variation show that the process is reversible:

```coq
Definition foo := 10.
Print foo. (* foo = 10 *)
Close Scope nat_scope.
Print foo. (* foo = 10%nat *)
Undelimit Scope nat_scope.
Print foo. (* foo = S 9 *)

Delimit Scope nat_scope with nat.
Print foo. (* foo = 10%nat *)
Open Scope nat_scope.
Print foo. (* foo = 10 *)
```

Reply:
Yes, it's reversible, and this is important.  Being able to temporarily disable the printing of notations in just one scope is important for debugging, and can also be important in customizing printing.  It's also useful in defining scopes in notations, as in https://github.com/coq/coq/issues/9751#issuecomment-472432207 .  It also shows up in https://github.com/coq/coq/issues/8805#issuecomment-432281939 .

Reply:
> Being able to temporarily disable the printing of notations in just one scope is important for debugging,

"scope" meaning a scope key, not a scope name, right?

Can you give a realistic example where it's used to debug a user error?  The issues you linked to are using scope keys to debug Coq itself, a less frequent use case.  Is it useful for debugging printing problems, parsing problems, or both?  I'm not seeing what that would be; I've not used the feature much.

I think this information would go into the section on printing or parsing.

Also, I notice that `Delimit Scope C_scope with C.` from #8805 gives the warning message

```
Declaring a scope implicitly is deprecated; use in advance an explicit
"Declare Scope C_scope.". [undeclared-scope,deprecated]
```

Perhaps it should not be (it seems harmless), and the doc should mention this if the deprecation is removed.  OTOH, it only saves the one line `Declare Scope C_scope`

@Zimmi48, your thoughts?

Reply:
> "scope" meaning a scope key, not a scope name, right?
> 
> Can you give a realistic example where it's used to debug a user error? The issues you linked to are using scope keys to debug Coq itself, a less frequent use case. Is it useful for debugging printing problems, parsing problems, or both? I'm not seeing what that would be; I've not used the feature much.

No, I mean a scope name.  (But currently there's a one-to-one-correspondence in the wild, I believe.)  It's useful for debugging unification issues when `Set Printing All` / `Unset Printing Notations` is too much of a hammer, and the relevant notations are just in a particular scope.

> Perhaps it should not be (it seems harmless), and the doc should mention this if the deprecation is removed. 

The justification is given in https://github.com/coq/coq/issues/7120 :
> `@RalfJung` commented (https://github.com/coq/coq/pull/6247#issuecomment-377202990) that the current implicit binding is awkward, since typos will silently create a new scope. I agree.

Reply:
> The issues you linked to are using scope keys to debug Coq itself, a less frequent use case. 

Not all of them.  There is a use-case in https://github.com/coq/coq/issues/9751#issuecomment-472432207 which involves the confluence of three things:
1. You have a scope `test_scope` for which you want printing-and-parsing notations
2. You want notations in `test_scope` to print *only* when the scope is open, or when it has been enabled via `Arguments` or special notations, and you want it to never show up in printing via `(...)%test`.
3. You want to issue an `Arguments` or `Notation` command which binds the scope to a particular argument of the function/notation, so that the scope is opened automatically for parsing and printing in that position.

Reply:
That may be a good abstract statement.  It's neither a simple nor short statement.  I hoped you could give a _concrete_ example that has a few commands and a specific problem (expected _this_ output but got _that_ output).  Then the writeup might be along the lines of:

"Undelimit Scope, ... can be helpful for debugging printing and parsing notations.  For example, this: <commands> produced A when the user expected B.  Adding <commands>, you can see that the problem was X."

Or I could leave it at:  "Undelimit Scope, ... can be used to turn scopes and scope delimiters on and off, which can be helpful in debugging printing and parsing notations."

For either of those, I'd need a little help getting the wording details right.

Or we could just leave it as a todo.

FWIW I'm going to get some exercise before the sun goes down.  Back in an hour or two.

Reply:
Here's a rather contrived and slightly lengthy but hopefully somewhat helpful example:
```coq
Axiom A : Set.
Axiom B : Set.
Axiom A_of_nat : nat -> A.
Axiom B_of_nat : nat -> B.
Axiom plusA : A -> A -> A.
Axiom plusB : B -> B -> B.
Delimit Scope A_scope with A.
Delimit Scope B_scope with B.
Coercion A_of_nat : nat >-> A.
Coercion B_of_nat : nat >-> B.
Infix "+" := plusA : A_scope.
Infix "+" := plusB : B_scope.
Axiom A_to_B : A -> B.
Arguments A_to_B _%A.
Coercion A_to_B : A >-> B.
Axiom P : forall {A}, A -> Prop.
Arguments P {_} _%B.
Notation P2 := P.
Delimit Scope foo_scope with foo.
Notation "'P2' x" := (P (A_to_B x)) (at level 10) : foo_scope.
Open Scope foo_scope.
Goal let x := 100 in
     let y := 101 in
     (P (plusB x y) -> P (plusB y x))
     -> (@P B (plusA x y) -> @P B (plusA y x)).
  cbv zeta; intro H.
  Fail apply H.
  (* Error: In environment
H : P2 (100 + 101) -> P2 (101 + 100)
Unable to unify "P2 (100 + 101) -> P2 (101 + 100)" with
 "P2 (100 + 101) -> P2 (101 + 100)".
   *)
  Set Printing All. (* good luck reading this with all the [S]s *)
  Unset Printing All.
  Arguments P : clear scopes.
  Close Scope foo_scope.
  Undelimit Scope foo_scope.
  Undelimit Scope B_scope.
  Undelimit Scope A_scope.
  Fail apply H.
  (* The command has indeed failed with message:
In environment
In environment
H : P2 (plusB 100 101) -> P2 (plusB 101 100)
Unable to unify "P2 (plusB 100 101) -> P2 (plusB 101 100)" with
 "P2 (plusA 100 101) -> P2 (plusA 101 100)".
*)
```

Maybe something like
"Undelimit Scope, ... can be used to turn off the printing of notations bound to a particular scope, which can be useful when notations in a given scope are getting in the way of understanding a goal, but turning off all notations would make the goal unreadable."
?

Reply:
Thanks.  OK, I see that first Fail shows `+` to represent two different things and the second one distinguishes them.  I think I can write a sentence or two on this.

FWIW, I've wanted to learn how coercions work and what they're good for (among other features), but the manual doesn't explain this.

Reply:
If you want to play around with coercions, I think the most useful identifiers to turn into coercions are is_true, Z.of_nat, Zpos, Npos, Z.of_N, and Q.inj_Z (I think that's what the Z -> Q function is called).

Reply:
I put this below the description of `Printing All`:

```
 .. note:: In some cases, setting :flag:`Printing All` may display terms
      that are so big they become very hard to read.  One technique to work around
      this is use :cmd:`Undelimit Scope` and/or :cmd:`Close Scope` to turn off the
      printing of notations bound to particular scope(s).  This can be useful when
      notations in a given scope are getting in the way of understanding
      a goal, but turning off all notations with :flag:`Printing All` would make
      the goal unreadable.
```

Reply:
That seems good, though I wonder if it might be better placed with `Printing Notations` rather than `Printing All|

Reply:
Yeah, I thought about that a little.  While the notations info is pretty understandable, introducing concepts step by step, example by example, it's more of a tutorial than a reference manual style where the full spec is at the beginning and each aspect is explained in turn.  So this note didn't seem to fit the style so well under "Displaying symbolic notations".

I've given zero thought to further improvements to the syntax extensions chapter.  And lots of things will be moving around going forward.  Plenty of opportunity to put it in a better spot later.